### DISCUSSION TITLE
libcURL to use the default system cert store

### DISCUSSION THREAD

**Discussion by pkcs11newbie** [curl/curl#11707] [Q&A]
Hello,

I'm using libCurl ( built to use OpenSSL ) in our C++ client application to connect to an Https based server. 

My client can connect fine to a server that's based of a certificate that's issued by a self-signed CA authority.  I set below option for CA cert to trust the server's certificate.
  CURL_SET_OPT( m_curlHandle, CURLOPT_CAINFO, m_ctxInfo.getCAPemCertPath().c_str() );

However, I like my same client application to connect to a server that are using certificate signed by well known trusted CA authority as well. TO achieve this, I realized that my client application needs to look into local system cert store ( windows, linux ):

Based on the documentation,
https://curl.se/libcurl/c/CURLOPT_CAINFO.html
https://curl.se/libcurl/c/CURLOPT_SSL_OPTIONS.html

I modified client application: 
-  To not set CURLOPT_CAINFO option.
-  Set option: CURL_SET_OPT( m_curlHandle, CURLOPT_SSL_OPTIONS, (long)CURLSSLOPT_NATIVE_CA );

Regardless of above modification,  I see error : CURL error: 60 message: SSL peer certificate or SSH remote key was not OK

I wasn't quite sure where it's looking for the cert store, I modified to see if I can get that information. Unfortunately, I got cainfo as NULL.
        char* cainfo = NULL;
        curl_easy_getinfo( m_curlHandle, CURLINFO_CAINFO, &cainfo );
        if ( cainfo )
        {
            printf( "default ca info path: %s\n", cainfo );
        }

I'm not quite sure what can I do so libCurl could autodetect the local cert store. Any help greatly appreciated.

Thank you.

**Comment by bagder** [comment]
> I'm using libCurl ( built to use OpenSSL ) in our C++ client application to connect to an Https based server.

`CURLSSLOPT_NATIVE_CA` only works on Windows if you built with OpenSSL. What libcurl version are you using?

On linux, the CA cert is typically loaded from a file and if you build with your system libcurl it will use that by default.

> Unfortunately, I got cainfo as NULL.

On Windows that is expected since OpenSSL has no default path there. On Linux, that is not typical.

**Comment by pkcs11newbie** [comment]
Thank you so much for your quick reply. I'm still learning on how to leverage libcurl in our application. Please do pardon my ignorance.

We built libcurl locally out here with OpenSSL. We are using 8.2.1 ( based on curlver.h )

On windows, CURLSSLOPT_NATIVE_CA option should have worked then. Is there a way for me to figure out or understand why didn't this option work?

I haven't tested on Linux yet, Just thinking out loud here:
As you mentioned, on Linux, I should get default path from curl_easy_getinfo( m_curlHandle, CURLINFO_CAINFO, &cainfo ). Given the default path, I can then set CURL_SET_OPT( m_curlHandle, CURLOPT_CAINFO, m_ctxInfo.getCAPemCertPath().c_str() );

Do I understand this correctly ?

Thank you.

**Reply by jay** [reply]
> Is there a way for me to figure out or understand why didn't this option work?

Turn on verbose mode [CURLOPT_VERBOSE](https://curl.se/libcurl/c/CURLOPT_VERBOSE.html) to get more information.

> As you mentioned, on Linux, I should get default path from curl_easy_getinfo( m_curlHandle, CURLINFO_CAINFO, &cainfo ).

I don't see where he mentioned that and there should be no reason for you to do that.

On Linux there's usually a hard coded certificate bundle or certificate path from when libcurl was built and it will use that as a default if you don't set the locations.

On Windows if libcurl is using the Schannel backend (Windows OS native SSL backend) then by default it uses the OS certificate store and CURLOPT_CAINFO can be used to specify a certificate bundle to override it. On Windows if libcurl is using other SSL backends then there may or may not be a default location. See [this](https://curl.se/docs/ssl-compared.html). You may be able to set CURLOPT_CAINFO and/or CURLOPT_CAPATH.

In addition, [CURLSSLOPT_NATIVE_CA](https://curl.se/libcurl/c/CURLOPT_SSL_OPTIONS.html) can be used in some cases.
